// ng-zorro-antd/esm5/date-picker/standard-types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function PickerResultSingle() { }
function PickerResultSingle_tsickle_Closure_declarations() {
    /** @type {?} */
    PickerResultSingle.prototype.date;
    /** @type {?} */
    PickerResultSingle.prototype.dateString;
}
/**
 * @record
 */
export function PickerResultRange() { }
function PickerResultRange_tsickle_Closure_declarations() {
    /** @type {?} */
    PickerResultRange.prototype.date;
    /** @type {?} */
    PickerResultRange.prototype.dateString;
}
/**
 * @record
 */
export function DisabledTimeConfig() { }
function DisabledTimeConfig_tsickle_Closure_declarations() {
    /** @type {?} */
    DisabledTimeConfig.prototype.nzDisabledHours;
    /** @type {?} */
    DisabledTimeConfig.prototype.nzDisabledMinutes;
    /** @type {?} */
    DisabledTimeConfig.prototype.nzDisabledSeconds;
}
/**
 * @record
 */
export function SupportTimeOptions() { }
function SupportTimeOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzFormat;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzHourStep;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzMinuteStep;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzSecondStep;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzDisabledHours;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzDisabledMinutes;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzDisabledSeconds;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzHideDisabledOptions;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzDefaultOpenValue;
    /** @type {?|undefined} */
    SupportTimeOptions.prototype.nzAddOn;
}
/**
 * @record
 */
export function PresetRanges() { }
function PresetRanges_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [ key: string ]: Date[];
    */
}
