// ng-zorro-antd/esm5/transfer/interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function TransferItem() { }
function TransferItem_tsickle_Closure_declarations() {
    /** @type {?} */
    TransferItem.prototype.title;
    /** @type {?|undefined} */
    TransferItem.prototype.direction;
    /** @type {?|undefined} */
    TransferItem.prototype.disabled;
    /** @type {?|undefined} */
    TransferItem.prototype.checked;
    /** @type {?|undefined} */
    TransferItem.prototype._hiden;
    /* TODO: handle strange member:
    [key: string]: any;
    */
}
/**
 * @record
 */
export function TransferCanMove() { }
function TransferCanMove_tsickle_Closure_declarations() {
    /** @type {?} */
    TransferCanMove.prototype.direction;
    /** @type {?} */
    TransferCanMove.prototype.list;
}
/**
 * @record
 */
export function TransferChange() { }
function TransferChange_tsickle_Closure_declarations() {
    /** @type {?} */
    TransferChange.prototype.from;
    /** @type {?} */
    TransferChange.prototype.to;
    /** @type {?} */
    TransferChange.prototype.list;
}
/**
 * @record
 */
export function TransferSearchChange() { }
function TransferSearchChange_tsickle_Closure_declarations() {
    /** @type {?} */
    TransferSearchChange.prototype.direction;
    /** @type {?} */
    TransferSearchChange.prototype.value;
}
/**
 * @record
 */
export function TransferSelectChange() { }
function TransferSelectChange_tsickle_Closure_declarations() {
    /** @type {?} */
    TransferSelectChange.prototype.direction;
    /** @type {?} */
    TransferSelectChange.prototype.checked;
    /** @type {?} */
    TransferSelectChange.prototype.list;
    /** @type {?} */
    TransferSelectChange.prototype.item;
}
