// ng-zorro-antd/esm2015/message/nz-message.definitions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function NzMessageDataOptions() { }
function NzMessageDataOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NzMessageDataOptions.prototype.nzDuration;
    /** @type {?|undefined} */
    NzMessageDataOptions.prototype.nzAnimate;
    /** @type {?|undefined} */
    NzMessageDataOptions.prototype.nzPauseOnHover;
}
/**
 * @record
 */
export function NzMessageData() { }
function NzMessageData_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NzMessageData.prototype.type;
    /** @type {?|undefined} */
    NzMessageData.prototype.content;
}
/**
 * @record
 */
export function NzMessageDataFilled() { }
function NzMessageDataFilled_tsickle_Closure_declarations() {
    /** @type {?} */
    NzMessageDataFilled.prototype.messageId;
    /** @type {?|undefined} */
    NzMessageDataFilled.prototype.state;
    /** @type {?|undefined} */
    NzMessageDataFilled.prototype.options;
    /** @type {?} */
    NzMessageDataFilled.prototype.createdAt;
}
