// ng-zorro-antd/esm5/modal/nz-modal.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 * @template T, R
 */
export function ModalOptions() { }
function ModalOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ModalOptions.prototype.nzModalType;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzVisible;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzZIndex;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzWidth;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzWrapClassName;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzClassName;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzStyle;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzIconType;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzTitle;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzContent;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzComponentParams;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzClosable;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzKeyboard;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzMask;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzMaskClosable;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzMaskStyle;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzBodyStyle;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzFooter;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzGetContainer;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzAfterOpen;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzAfterClose;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzOkText;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzOkType;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzOkLoading;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzOnOk;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzCancelText;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzCancelLoading;
    /** @type {?|undefined} */
    ModalOptions.prototype.nzOnCancel;
}
/**
 * @record
 * @template T
 */
export function ModalOptionsForService() { }
function ModalOptionsForService_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ModalOptionsForService.prototype.nzOnOk;
    /** @type {?|undefined} */
    ModalOptionsForService.prototype.nzOnCancel;
}
/**
 * @record
 * @template T
 */
export function ModalButtonOptions() { }
function ModalButtonOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    ModalButtonOptions.prototype.label;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.type;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.shape;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.ghost;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.size;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.autoLoading;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.show;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.loading;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.disabled;
    /** @type {?|undefined} */
    ModalButtonOptions.prototype.onClick;
}
